Running Vaadin 14 project in npm mode decides to skip npm install and fails because NPE from getStatsFromWebpack

Vaadin 14 rc3. 
Spring project created with Eclipse plugin preview. Eclipse v2019 latest.

Project works first, but at some point due to errors I removed a few files and node_modules folder. Debug as Java Application to the Spring Boot project starts the app (server), but does not work due to NPE at com.vaadin.flow.server.frontend.FrontendUtils.getStatsFromWebpack(FrontendUtils.java:349) ~[flow-server-2.0.2.jar:2.0.2]

I can observe that node_modules does not contain much anything. In the log I see the build skips npm install. Removing target, package.json, package-lock.json and webpack.*.js and running the project again now starts npm install and my node_modules is again 17k files.

[rc3-spring-fail.txt](https://github.com/vaadin/flow/files/3358978/rc3-spring-fail.txt)

I tried to reproduce it using Eclipse plugin preview. After creating the project, I ran `mvn install` and I got this error:
```
  Failed to find the following imports in the `node_modules` tree:
      ➜ @vaadin/flow-frontend/flow-component-renderer.js
  If the build fails, check that npm packages are installed.

[INFO] Updated /Users/mehdi/eclipse-workspace/my-starter-project/target/frontend/generated-flow-imports.js
[INFO] Running webpack ...
Hash: 677c0d156d1d5d33c613
Version: webpack 4.30.0
Time: 7157ms
Built at: 07/09/2019 11:08:58 AM
 2 assets
Entrypoint bundle = build/vaadin-bundle-4334868b8893557fa814.cache.js
Entrypoint bundle.es5 = build/vaadin-bundle.es5-aceb16b7bde3950865bc.cache.js
[18] ../target/frontend/generated-flow-imports.js?babel-target=es5 + 83 modules 745 KiB {1} [built]
     | ../target/frontend/generated-flow-imports.js?babel-target=es5 770 bytes [built]
     |     + 83 hidden modules
[19] ../target/frontend/generated-flow-imports.js?babel-target=es6 + 83 modules 662 KiB {0} [built]
     | ../target/frontend/generated-flow-imports.js?babel-target=es6 772 bytes [built]
     |     + 83 hidden modules
    + 18 hidden modules

ERROR in ../target/frontend/generated-flow-imports.js?babel-target=es5
Module not found: Error: Can't resolve '@vaadin/flow-frontend/flow-component-renderer.js?babel-target=es5' in '/Users/mehdi/eclipse-workspace/my-starter-project/target/frontend'
 @ ../target/frontend/generated-flow-imports.js?babel-target=es5 13:0-58

ERROR in ../target/frontend/generated-flow-imports.js?babel-target=es6
Module not found: Error: Can't resolve '@vaadin/flow-frontend/flow-component-renderer.js?babel-target=es6' in '/Users/mehdi/eclipse-workspace/my-starter-project/target/frontend'
 @ ../target/frontend/generated-flow-imports.js?babel-target=es6 13:0-58
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 34.910 s
[INFO] Finished at: 2019-07-09T11:08:58+03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.vaadin:vaadin-maven-plugin:14.0.0.rc3:build-frontend (default) on project my-starter-project: Execution default of goal com.vaadin:vaadin-maven-plugin:14.0.0.rc3:build-frontend failed: Webpack process exited with non-zero exit code.
[ERROR] Stderr: ''
```

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/NodeUpdater.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/NodeUpdater.java
@@ -271,7 +271,7 @@ public abstract class NodeUpdater implements FallibleCommand {
                 UTF_8.name());
     }
 
-    static Logger log() {
+    Logger log() {
         // Using short prefix so as npm output is more readable
         return LoggerFactory.getLogger("dev-updater");
     }

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -30,6 +30,9 @@ import static com.vaadin.flow.server.frontend.FrontendUtils.FLOW_NPM_PACKAGE_NAM
  */
 public class TaskRunNpmInstall implements FallibleCommand {
 
+    static final String SKIPPING_NPM_INSTALL = "Skipping `npm install`.";
+    static final String RUNNING_NPM_INSTALL = "Running `npm install` ...";
+
     private final NodeUpdater packageUpdater;
 
     /**
@@ -46,10 +49,10 @@ public class TaskRunNpmInstall implements FallibleCommand {
     @Override
     public void execute() throws ExecutionFailedException {
         if (packageUpdater.modified || shouldRunNpmInstall()) {
-            packageUpdater.log().info("Running `npm install` ...");
+            packageUpdater.log().info(RUNNING_NPM_INSTALL);
             runNpmInstall();
         } else {
-            packageUpdater.log().info("Skipping `npm install`.");
+            packageUpdater.log().info(SKIPPING_NPM_INSTALL);
         }
     }
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskRunNpmInstall.java
@@ -24,7 +24,6 @@ import com.vaadin.flow.server.ExecutionFailedException;
 import com.vaadin.flow.server.FallibleCommand;
 
 import static com.vaadin.flow.server.frontend.FrontendUtils.FLOW_NPM_PACKAGE_NAME;
-import static com.vaadin.flow.server.frontend.NodeUpdater.log;
 
 /**
  * Run <code>npm install</code> after dependencies have been updated.
@@ -47,10 +46,10 @@ public class TaskRunNpmInstall implements FallibleCommand {
     @Override
     public void execute() throws ExecutionFailedException {
         if (packageUpdater.modified || shouldRunNpmInstall()) {
-            log().info("Running `npm install` ...");
+            packageUpdater.log().info("Running `npm install` ...");
             runNpmInstall();
         } else {
-            log().info("Skipping `npm install`.");
+            packageUpdater.log().info("Skipping `npm install`.");
         }
     }
 
@@ -58,7 +57,8 @@ public class TaskRunNpmInstall implements FallibleCommand {
         if (packageUpdater.nodeModulesFolder.isDirectory()) {
             File[] installedPackages = packageUpdater.nodeModulesFolder
                     .listFiles();
-            return installedPackages == null
+            assert installedPackages != null;
+            return installedPackages.length == 0
                     || (installedPackages.length == 1 && FLOW_NPM_PACKAGE_NAME
                             .startsWith(installedPackages[0].getName()));
         }
@@ -82,17 +82,17 @@ public class TaskRunNpmInstall implements FallibleCommand {
             process = builder.inheritIO().start();
             int errorCode = process.waitFor();
             if (errorCode != 0) {
-                log().error(
+                packageUpdater.log().error(
                         ">>> Dependency ERROR. Check that all required dependencies are deployed in npm repositories.");
                 throw new ExecutionFailedException(
                         "Npm install has exited with non zero status. "
                                 + "Some dependencies are not installed. Check npm command output");
             } else {
-                log().info(
+                packageUpdater.log().info(
                         "package.json updated and npm dependencies installed. ");
             }
         } catch (InterruptedException | IOException e) {
-            log().error("Error when running `npm install`", e);
+            packageUpdater.log().error("Error when running `npm install`", e);
             throw new ExecutionFailedException(
                     "Command 'npm install' failed to finish", e);
         } finally {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskUpdateWebpack.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/TaskUpdateWebpack.java
@@ -24,12 +24,13 @@ import java.nio.file.Path;
 import java.util.List;
 
 import org.apache.commons.io.FileUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import com.vaadin.flow.server.FallibleCommand;
 
 import static com.vaadin.flow.server.frontend.FrontendUtils.WEBPACK_CONFIG;
 import static com.vaadin.flow.server.frontend.FrontendUtils.WEBPACK_GENERATED;
-import static com.vaadin.flow.server.frontend.NodeUpdater.log;
 
 /**
  * Updates the webpack config file according with current project settings.
@@ -144,5 +145,10 @@ public class TaskUpdateWebpack implements FallibleCommand {
                 ? webpackConfigPath.relativize(path)
                 : path;
         return relativePath.toString().replaceAll("\\\\", "/");
+    }
+
+    private Logger log() {
+        // Using short prefix so as npm output is more readable
+        return LoggerFactory.getLogger("dev-updater");
     }
 }

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
@@ -73,7 +73,7 @@ public class TaskRunNpmInstallTest {
         nodeUpdater.modified = false;
         task.execute();
 
-        Mockito.verify(logger).info("Running `npm install` ...");
+        Mockito.verify(logger).info(TaskRunNpmInstall.RUNNING_NPM_INSTALL);
     }
 
     @Test
@@ -85,7 +85,7 @@ public class TaskRunNpmInstallTest {
         nodeUpdater.modified = false;
         task.execute();
 
-        Mockito.verify(logger).info("Skipping `npm install`.");
+        Mockito.verify(logger).info(TaskRunNpmInstall.SKIPPING_NPM_INSTALL);
     }
 
     @Test
@@ -97,7 +97,7 @@ public class TaskRunNpmInstallTest {
         nodeUpdater.modified = false;
         task.execute();
 
-        Mockito.verify(logger).info("Running `npm install` ...");
+        Mockito.verify(logger).info(TaskRunNpmInstall.RUNNING_NPM_INSTALL);
     }
 
     @Test
@@ -106,6 +106,6 @@ public class TaskRunNpmInstallTest {
         nodeUpdater.modified = true;
         task.execute();
 
-        Mockito.verify(logger).info("Running `npm install` ...");
+        Mockito.verify(logger).info(TaskRunNpmInstall.RUNNING_NPM_INSTALL);
     }
 }

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
@@ -25,6 +25,7 @@ import org.junit.rules.TemporaryFolder;
 import org.mockito.Mockito;
 import org.slf4j.Logger;
 
+import com.vaadin.flow.server.ExecutionFailedException;
 import com.vaadin.flow.server.frontend.scanner.ClassFinder;
 import com.vaadin.flow.server.frontend.scanner.FrontendDependencies;
 
@@ -65,7 +66,8 @@ public class TaskRunNpmInstallTest {
     }
 
     @Test
-    public void runNpmInstall_emptyDir_npmInstallIsExecuted() {
+    public void runNpmInstall_emptyDir_npmInstallIsExecuted()
+            throws ExecutionFailedException {
         File nodeModules = new File(npmFolder, NODE_MODULES);
         nodeModules.mkdir();
         nodeUpdater.modified = false;
@@ -76,7 +78,7 @@ public class TaskRunNpmInstallTest {
 
     @Test
     public void runNpmInstall_nonEmptyDir_npmInstallIsNotExecuted()
-            throws IOException {
+            throws IOException, ExecutionFailedException {
         File nodeModules = new File(npmFolder, NODE_MODULES);
         nodeModules.mkdir();
         new File(nodeModules, "foo").createNewFile();
@@ -88,7 +90,7 @@ public class TaskRunNpmInstallTest {
 
     @Test
     public void runNpmInstall_dirContainsOnlyFlowNpmPackage_npmInstallIsNotExecuted()
-            throws IOException {
+            throws IOException, ExecutionFailedException {
         File nodeModules = new File(npmFolder, NODE_MODULES);
         nodeModules.mkdir();
         new File(nodeModules, "@vaadin/flow-frontend/").mkdirs();
@@ -99,7 +101,8 @@ public class TaskRunNpmInstallTest {
     }
 
     @Test
-    public void runNpmInstall_modified_npmInstallIsExecuted() {
+    public void runNpmInstall_modified_npmInstallIsExecuted()
+            throws ExecutionFailedException {
         nodeUpdater.modified = true;
         task.execute();
 

--- a/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
+++ b/flow-server/src/test/java/com/vaadin/flow/server/frontend/TaskRunNpmInstallTest.java
@@ -1,0 +1,108 @@
+/*
+ * Copyright 2000-2018 Vaadin Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.server.frontend;
+
+import java.io.File;
+import java.io.IOException;
+
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.mockito.Mockito;
+import org.slf4j.Logger;
+
+import com.vaadin.flow.server.frontend.scanner.ClassFinder;
+import com.vaadin.flow.server.frontend.scanner.FrontendDependencies;
+
+import static com.vaadin.flow.server.frontend.FrontendUtils.NODE_MODULES;
+
+public class TaskRunNpmInstallTest {
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    private NodeUpdater nodeUpdater;
+
+    private TaskRunNpmInstall task;
+
+    private File npmFolder;
+
+    private Logger logger = Mockito.mock(Logger.class);
+
+    @Before
+    public void setUp() throws IOException {
+        npmFolder = temporaryFolder.newFolder();
+        nodeUpdater = new NodeUpdater(Mockito.mock(ClassFinder.class),
+                Mockito.mock(FrontendDependencies.class), npmFolder,
+                new File("")) {
+
+            @Override
+            public void execute() {
+            }
+
+            @Override
+            Logger log() {
+                return logger;
+            }
+
+        };
+        task = new TaskRunNpmInstall(nodeUpdater);
+
+    }
+
+    @Test
+    public void runNpmInstall_emptyDir_npmInstallIsExecuted() {
+        File nodeModules = new File(npmFolder, NODE_MODULES);
+        nodeModules.mkdir();
+        nodeUpdater.modified = false;
+        task.execute();
+
+        Mockito.verify(logger).info("Running `npm install` ...");
+    }
+
+    @Test
+    public void runNpmInstall_nonEmptyDir_npmInstallIsNotExecuted()
+            throws IOException {
+        File nodeModules = new File(npmFolder, NODE_MODULES);
+        nodeModules.mkdir();
+        new File(nodeModules, "foo").createNewFile();
+        nodeUpdater.modified = false;
+        task.execute();
+
+        Mockito.verify(logger).info("Skipping `npm install`.");
+    }
+
+    @Test
+    public void runNpmInstall_dirContainsOnlyFlowNpmPackage_npmInstallIsNotExecuted()
+            throws IOException {
+        File nodeModules = new File(npmFolder, NODE_MODULES);
+        nodeModules.mkdir();
+        new File(nodeModules, "@vaadin/flow-frontend/").mkdirs();
+        nodeUpdater.modified = false;
+        task.execute();
+
+        Mockito.verify(logger).info("Running `npm install` ...");
+    }
+
+    @Test
+    public void runNpmInstall_modified_npmInstallIsExecuted() {
+        nodeUpdater.modified = true;
+        task.execute();
+
+        Mockito.verify(logger).info("Running `npm install` ...");
+    }
+}

